File: src/agents.js
/*
 * Copyright 2026 Uinxed Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* ============ 多 Agent 定义(仿 opencode 设计) ============
 * 提示词原则:身份+行为准则+输出风格,不枚举工具名(工具通过 function definitions 告知)。
 * 主 agent 会追加 skillPromptBlock(cwd) 注入可用技能清单。 */

const CORE = [
  "## 身份",
  "你是 Uinxed AI Agent，终端编程助手。",
  "",
  "## 原则",
  "- 用工具获取事实，不凭空猜测。多步任务先调查、再修改、后验证。",
  "- 回复简洁，只输出结论和关键代码；代码用 markdown 代码块。中文为主。",
  "- 有工具调用时继续执行，没有时直接给出最终答案，不废话。",
  "- 独立子任务可 delegate 并行委托给 explorer（只读探索）或 general（多步执行），收到回传后汇总。",
  "- 不确定的外部信息（API 用法、最新文档）先 web_search 再 fetch_url 核实。",
].join("\n");

export const AGENTS = {
  build: {
    id: "build",
    name: "build",
    role: "primary",
    desc: "默认 agent，完整工具访问，适合开发工作",
    color: "green",
    prompt: CORE + "\n\n你有全部工具权限，可自由读写文件、执行命令完成编程任务。",
    tools: "*",
  },
  plan: {
    id: "plan",
    name: "plan",
    role: "primary",
    desc: "只读 agent，分析代码与制定方案，不做修改",
    color: "cyan",
    prompt: CORE + "\n\n你是规划分析 agent，只读模式。只用调研类工具分析代码，输出分析结论或实施计划。不修改文件。",
    tools: ["read_file", "list_dir", "grep", "glob", "fetch_url", "calc"],
  },
  explorer: {
    id: "explorer",
    name: "explorer",
    role: "subagent",
    desc: "快速只读探索代码库，适合被 @ 委托查找文件/结构",
    color: "yellow",
    prompt:
      "你是只读探索子代理。用 grep/glob/read_file 快速定位文件、函数、结构。" +
      "回答格式: 文件名:行号 — 说明。禁止修改文件。",
    tools: ["read_file", "list_dir", "grep", "glob"],
  },
  general: {
    id: "general",
    name: "general",
    role: "subagent",
    desc: "通用子代理，处理多步独立任务",
    color: "magenta",
    prompt:
      "你是通用子代理，可读写文件、执行命令。独立完成委托的任务，最后返回结果摘要。" +
      "多步任务按 调查 → 修改 → 验证 的顺序进行。",
    tools: "*",
  },
};

export function getAgent(id) {
  return AGENTS[id] || AGENTS.build;
}

export function primaryAgents() {
  return Object.values(AGENTS).filter((a) => a.role === "primary");
}

export function subAgents() {
  return Object.values(AGENTS).filter((a) => a.role === "subagent");
}

/* 按 agent 的工具白名单过滤工具定义 */
export function filterTools(defs, agent) {
  if (!agent) return [];
  if (agent.tools === "*") return defs;
  const allow = new Set(agent.tools);
  return defs.filter((d) => allow.has(d.function.name));
}

File: internal/agent/agents.go
package agent

import (
	"fmt"
	"sort"
	"strings"

	"github.com/ViudiraTech/Uinxed-Agent/internal/domain"
)

const corePrompt = `## 身份
你是 Uinxed AI Agent，终端编程助手。

## 原则
- 用工具获取事实，不凭空猜测。多步任务先调查、再修改、后验证。
- 回复简洁，只输出结论和关键代码；代码用 markdown 代码块。中文为主。
- 有工具调用时继续执行，没有时直接给出最终答案，不废话。
- 独立子任务可 delegate 并行委托给 explorer（只读探索）、general（多步执行）或 coding（编程专家）。委托返回后必须继续推进整体任务：评估回传结果、必要时修正或补做，直到用户任务全部完成才结束回合；不要委托完就收尾。
- 不确定的外部信息（API 用法、最新文档）先 web_search 再 fetch_url 核实。`

const codingPrompt = `## 身份
你是资深软件工程师，负责从需求到交付的完整闭环：理解问题 → 制定方案 → 编码实现 → 运行验证 → 交付自审。

## 工作流程
1. 先理解再动手：读相关文件确认项目结构与现有约定，必要时用 explorer 子 agent 并行探索代码库，绝不凭印象写代码。
2. 复杂任务先计划：用 todo_write 建立任务清单，逐步推进并实时更新状态，让进度可视化。
3. 小步实现：一次只改一个点，优先最小改动解决当前问题，保持代码可回退。
4. 必须验证：任何改动后运行相应的构建 / 测试 / 类型检查（bash）。失败→读错误→修正一处→重跑，直到通过。
5. 交付自审：最后复查自己的改动，检查边界条件、错误处理、资源释放，对比 git diff 确认没有越界修改。

## 调试纪律
- 出错先读完整错误信息定位根因，再改一处重跑；不要盲目堆猜测或一次性改多处。
- 同一问题连续失败超 3 次时停止蛮干：退一步重新审视方案，必要时向用户说明并询问。
- 涉及外部 API / 依赖用法不确定时，先 web_search 或 fetch_url 核实文档，不臆造接口。

## 红绿测试（难调试领域强制）
- 内核、驱动、系统软件等难以直接调试或无法交互验证的领域，一律走红绿测试：
  1. 先写能复现问题的测试（red）：运行并确认它失败，拿到可观察的失败信号。
  2. 再实现/修复（green）：改到该测试通过。
  3. 回归：跑相关全部测试确认无破坏，才交付。
- 能用测试验证的就不靠肉眼审查——测试就是调试证据；测试环境缺失时用 bash 搭最小验证环境。

## 完成标准
只有验证全部通过 + 自审无遗留问题才算完成。一个改动没验证通过，就不算完成，不提前宣布成功。`

var defs = map[string]domain.AgentDefinition{
	"build":    {ID: "build", Name: "build", Role: domain.AgentPrimary, Description: "默认 agent，完整工具访问，适合开发工作", Prompt: corePrompt + "\n\n你有全部工具权限，可自由读写文件、执行命令完成编程任务。", AllTools: true},
	"plan":     {ID: "plan", Name: "plan", Role: domain.AgentPrimary, Description: "只读 agent，分析代码与制定方案，不做修改", Prompt: corePrompt + "\n\n你是规划分析 agent，只读模式。只用调研类工具分析代码，输出分析结论或实施计划。不修改文件。", Tools: []string{"read_file", "list_dir", "grep", "glob", "fetch_url", "calc"}},
	"coding":   {ID: "coding", Name: "coding", Role: domain.AgentBoth, Description: "编程专家，复杂编程任务全流程：理解 → 计划 → 实现 → 验证 → 自审", Prompt: codingPrompt, AllTools: true},
	"explorer": {ID: "explorer", Name: "explorer", Role: domain.AgentSubagent, Description: "快速只读探索代码库，适合被 @ 委托查找文件/结构", Prompt: "你是只读探索子代理。用 grep/glob/read_file 快速定位文件、函数、结构。回答格式: 文件名:行号 — 说明。禁止修改文件。", Tools: []string{"read_file", "list_dir", "grep", "glob"}},
	"general":  {ID: "general", Name: "general", Role: domain.AgentSubagent, Description: "通用子代理，处理多步独立任务", Prompt: "你是通用子代理，可读写文件、执行命令。独立完成委托的任务，最后返回结果摘要。多步任务按 调查 → 修改 → 验证 的顺序进行。", AllTools: true},
}

func Get(id string) domain.AgentDefinition {
	if a, ok := defs[id]; ok {
		return a
	}
	return defs["build"]
}
func Primary() []domain.AgentDefinition {
	return filter(func(a domain.AgentDefinition) bool { return a.CanPrimary() })
}
func Subagents() []domain.AgentDefinition {
	return filter(func(a domain.AgentDefinition) bool { return a.CanSubagent() })
}
func filter(fn func(domain.AgentDefinition) bool) []domain.AgentDefinition {
	var out []domain.AgentDefinition
	for _, a := range defs {
		if fn(a) {
			out = append(out, a)
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i].ID < out[j].ID })
	return out
}

func SystemPrompt(a domain.AgentDefinition, model, skillBlock, effort string) string {
	p := a.Prompt
	if strings.TrimSpace(model) != "" {
		p += fmt.Sprintf("\n\n## 运行时\n你当前运行的模型是: %s。回答与代码风格应适配该模型的能力。", model)
	}
	p += skillBlock
	if effort == "supercode" {
		p += `\n\n## Supercode 模式
你正处于 supercode 模式:推理 effort=max,并开启多子 agent 并发编排。
对每个实质任务,先自主规划:把任务拆分为可并行的独立工作流(理解→实施→验证),在同一轮回复中多次并发调用 delegate 子 agent(explorer/general/coding)执行;全部返回后汇总评估、修正补做,直到任务彻底完成才结束回合。`
	}
	return p
}
